middleware.go

```diff
@@ -106,6 +117,98 @@ func createMiddleware(mw TykMiddleware) func(http.Handler) http.Handler {
+		return
+	}
+
+	if !mw.Base().Spec.CORS.OptionsPassthrough {
+		log.Debug("options passthrough is false - returning")
+		return
+	}
+
+	if r.Header.Get("Origin") == "" {
+		log.Debug("no origin header set in request so not a browser - returning")
+		return
+	}
+
+	// we don't need the request body for a CORS request
+	scheme := "http"
+	if config.Global.HttpServerOptions.UseSSL {
+		scheme += "s"
+	}
+
+	u := fmt.Sprintf("%s://localhost:%d/%s", scheme, config.Global.ListenPort, r.URL.Path)
+	corsRequest, err := http.NewRequest(http.MethodOptions, u, nil)
+
+	// https://www.w3.org/TR/2014/REC-cors-20140116/#cross-origin-request-with-preflight-0
+
+	corsRequest.Header.Set("Content-Type", r.Header.Get("Content-Type"))
+	corsRequest.Header.Set(request.Origin, r.Header.Get("Origin"))
+	corsRequest.Header.Set(request.AccessControlRequestMethod, r.Method)
+
+	// If author request headers is not empty include an Access-Control-Request-Headers header with as header field
+	// value a comma-separated list of the header field names from author request headers in lexicographical order,
+	// each converted to ASCII lowercase (even when one or more are a simple header).
+	var requestHeaders []string
+	for key := range r.Header {
+		requestHeaders = append(requestHeaders, strings.ToLower(key))
+	}
+	sort.Strings(requestHeaders)
+
+	corsRequest.Header.Set(request.AccessControlRequestHeaders, strings.Join(requestHeaders, ", "))
+
+	// TODO: Exclude Credentials
+	// The term user credentials for the purposes of this specification means cookies, HTTP authentication, and
+	// client-side SSL certificates that would be sent based on the user agent's previous interactions with the
+	// origin. Specifically it does not refer to proxy authentication or the Origin header. [COOKIES]
+
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	dump, err := httputil.DumpRequest(corsRequest, true)
+	if err != nil {
+		http.Error(w, fmt.Sprint(err), http.StatusInternalServerError)
+		return
+	}
+
+	log.Debugf("loopback request dump\n%s", string(dump))
+
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: config.Global.HttpServerOptions.SSLInsecureSkipVerify},
+	}
+
+	corsClient := &http.Client{Transport: tr}
+	corsRes, err := corsClient.Do(corsRequest)
+	if err != nil {
+		log.WithError(err).Error("Error from upstream OPTIONS request, ignoring and returning default response")
+	}
+
+	corsHeaders := map[string]string{}
+	corsHeaders[response.AccessControlAllowOrigin] = corsRes.Header.Get(response.AccessControlAllowOrigin)
+	corsHeaders[response.AccessControlAllowMethods] = corsRes.Header.Get(response.AccessControlAllowMethods)
+	corsHeaders[response.AccessControlAllowCredentials] = corsRes.Header.Get(response.AccessControlAllowCredentials)
+	corsHeaders[response.AccessControlMaxAge] = corsRes.Header.Get(response.AccessControlMaxAge)
+
+	for k, v := range corsHeaders {
+		if v != "" {
+			w.Header().Set(k, v)
```

Comment:
we need to make sure that response wasn't already sent to browser at this point. setting headers on `ttp.ResponseWriter` when response already sent won't take any affect